### PR TITLE
added Hw 2

### PR DESCRIPTION
**Pull request by maxkvant**

### CODE REVIEW

File: 2/homework/Makefile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+init:
+    pip install -r requirements.txt
+
+test:
+    py.test tests
+
+.PHONY: init test
```
Comment (Rauf-Kurbanov) [inline 2/homework/Makefile]: It is easier to maintain putting .PHONY right before the target

**Review by Rauf-Kurbanov — CHANGES_REQUESTED** [review]
Nice job, a lot of plots.

Common notice: 
1) no eof newlines
2) low f1 score alongside with high auc suggests poorly selected threshold or imbalanced validation set. 

To make this possible to review:
1) please eliminate all unused code from PR
2) Split your PR in two: in first merge all copy-pasted files [no need to review], in second: add your work on top of this so I can review it

**Comment by maxkvant** [comment]
> 1. no eof newlines

a new commit for it

> 2. Split your PR in two: in first merge all copy-pasted files [no need to review], in second: add your work on top of this so I can review it

just merged copy-pasted files into master

**Comment by Rauf-Kurbanov** [comment]
@maxkvant as a general comment please re-request review not only in email but also on a github

File: 2/homework/laughter_classification/sspnet_data_sampler.py [hunk 2]
```diff
@@ -88,14 +108,27 @@ def create_sampled_df(self, frame_sec, naudio=None, save_path=None, force_save=F
...
+        fullpaths = list(sorted(fullpaths))
 
-        colnames = ["V{}".format(i) for i in range(df.shape[1] - 2)]
-        colnames.append("IS_LAUGHTER")
-        colnames.append("SNAME")
-        df.columns = colnames
+        start_time = time.time()
+
+        def read_dataframe(self_iter_path_frame_n):
+            self_, iter, wav_path, frame_sec, n = self_iter_path_frame_n
```
Comment (Rauf-Kurbanov) [inline 2/homework/laughter_classification/sspnet_data_sampler.py]: `iter` is bad variable name

File: 2/homework/laughter_classification/sspnet_data_sampler.py [hunk 3]
```diff
@@ -88,14 +108,27 @@ def create_sampled_df(self, frame_sec, naudio=None, save_path=None, force_save=F
...
         fullpaths = self.get_valid_wav_paths()[:naudio]
-        dataframes = [self.df_from_file(wav_path, frame_sec) for wav_path in fullpaths]
-        df = pd.concat(dataframes)
+        fullpaths = list(sorted(fullpaths))
 
-        colnames = ["V{}".format(i) for i in range(df.shape[1] - 2)]
-        colnames.append("IS_LAUGHTER")
-        colnames.append("SNAME")
-        df.columns = colnames
+        start_time = time.time()
```
Comment (Rauf-Kurbanov) [inline 2/homework/laughter_classification/sspnet_data_sampler.py:117]: You might want to move start_time into `read_dataframe ` funcion fro more consistency
Another way to do this might be using a context manager
https://preshing.com/20110924/timing-your-code-using-pythons-with-statement/

File: 2/homework/laughter_classification/sspnet_data_sampler.py [hunk 4]
```diff
@@ -88,14 +108,27 @@ def create_sampled_df(self, frame_sec, naudio=None, save_path=None, force_save=F
...
+
+        def read_dataframe(self_iter_path_frame_n):
+            self_, iter, wav_path, frame_sec, n = self_iter_path_frame_n
+            res = self_.df_fbank_mfcc_from_file(wav_path, frame_sec)
+
+            if (iter > 2) and (iter & (iter - 1) == 0):
+                print("iter {}/{} {}".format(iter, n, time.time() - start_time))
```
Comment (Rauf-Kurbanov) [inline 2/homework/laughter_classification/sspnet_data_sampler.py]: Explore f-strings
https://www.python.org/dev/peps/pep-0498/

File: 2/homework/laughter_prediction/rnn.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        self.hidden_size = hidden_size
+        self.batch_size = batch_size
+
+        self.fbank_lstm = nn.LSTM(fbank_features, hidden_size)
+        self.mfcc_lstm  = nn.LSTM(mfcc_features, hidden_size)
+        self.linear_both = nn.Linear(hidden_size * 2, 2)
+        self.linear_mfcc = nn.Linear(hidden_size, 2)
+
+    def init_hidden_state(self, device):
+        return (torch.randn(1, self.batch_size, self.hidden_size).to(device),
```
Comment (Rauf-Kurbanov) [inline 2/homework/laughter_prediction/rnn.py:21]: In pytorch you don't need to specify device for particular layers since you can move your whole nn.Module to device all at once

File: 2/homework/laughter_prediction/predictors.py [hunk 6]
```diff
@@ -74,9 +77,18 @@ def __init__(self, threshold=0.045985743):
...
+        res = np.zeros_like(P, dtype=np.int)
+        res[P < threshold] = 0
+        res[P >= threshold] = 1
+        return res
+
+    def predict_proba(self, X, device='cpu'):
+        self.model.set_batch_size(len(X))
+        X_tensor = torch.tensor(X).float().to(device)
+        predicted_proba, _ = self.model.forward(X_tensor, device)
+        return torch.exp(predicted_proba[:,:,1]).cpu().detach().numpy()
```
Comment (Rauf-Kurbanov) [inline 2/homework/laughter_prediction/predictors.py:94]: I realise this might suggested by my poor choice of predictor interface. Just keep in mind that in pytorch it is ok to return logits logarithms without exponentiating. Also preserves numeric stability

File: 2/homework/laughter_prediction/rnn.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+                torch.randn(1, self.batch_size, self.hidden_size).to(device))
+
+    def forward(self, x, device='cpu'):
+        x_fbank = x[:, :, :self.fbank_features].permute(1, 0, 2)
+        x_mfcc  = x[:, :, self.fbank_features:].permute(1, 0, 2)
+
+        out_fbank_lstm, _ = self.fbank_lstm(x_fbank, self.init_hidden_state(device))
```
Comment (Rauf-Kurbanov) [inline 2/homework/laughter_prediction/rnn.py:28]: @maxkvant Why do you always initialize lstm hidden state with random value when you should pass the hidden from previous pass?

**Review by Rauf-Kurbanov — CHANGES_REQUESTED** [review]
General comments:
* Some minor violations of pep8
* Do not upload your checkpoints in repo, since it will become huge. Also take into account that you will have to carry this files in history even after removal. As an alternative you can wget from dropbox or use git-lfs.
* Left some minor suggestions on a code

On the notebook:
* You load `autoreload` plugin but don't use it
* You can add `;` after plot function to avoid unnecessary output
* Please do normalise you features !
* Adam is more safe optimisator choice if you are not planning to spend a lot of time on fine-tuning and lr-scheduling
* For general metrics you might want to use `sklearn.metric`
* Why would you need `global` in jupyter?
* Tensorboard for metrics is much better practice than pyplot. It also take care of smoothing
* Too much code in jupyter, it is ok for a homework but it is hard to review and maintain in real life.

Summary: please answer my question about hidden state and fix feature normalisation.
Other comments are just for you, can ignore them if you wish

**Comment by maxkvant** [comment]
> * Some minor violations of pep8

I didn't even know about pep8. I will read it. 

> * Do not upload your checkpoints in repo, since it will become huge. Also take into account that you will have to carry this files in history even after removal. As an alternative you can wget from dropbox or use 
git-lfs.

What do you consider as checkpoints? Binary files for models?

> On the notebook:
> 
> * You load `autoreload` plugin but don't use it

removed that line

> * Please do normalise you features !

added normalization

> * Adam is more safe optimisator choice if you are not planning to spend a lot of time on fine-tuning and lr-scheduling

Adam converges faster, but it overfits faster too. Now, I've switched to Adam, and I wasn't able to reproduce previous accuracy. 

> * For general metrics you might want to use `sklearn.metric`

Yes. I use it except for f1. 
I didn't figure out how to compute f1 for laughter class with function from `sklearn.metric`.

> * Why would you need `global` in jupyter?

Surprisingly, It crashes without `global`.

File: 2/homework/laughter_prediction/rnn.py (hunk 7 above)
Comment (maxkvant) [reply 2/homework/laughter_prediction/rnn.py:28]: I know 4 options for lstm initializartion:
1. zero state
2. random state
3. state from previous pass
4. noised state from previous pass

I have doubts about how to do approach 3 correctly for batches (Should I average it or not?). That's why I didn't consider options 3 and 4.

According to [post](https://r2rt.com/non-zero-initial-states-for-recurrent-neural-networks.html) random state better than zero state.

File: 2/homework/laughter_classification/sspnet_data_sampler.py (hunk 3 above)
Comment (maxkvant) [reply 2/homework/laughter_classification/sspnet_data_sampler.py:117]: `read_dataframe` is a helper for parallel feature computing.
I don't want to move it, because start time means the time before start of parallel computing.

**Comment by maxkvant** [comment]
Did I re-request review on github?

**Review by Rauf-Kurbanov — APPROVED** [review]
Approved, thanks 4/4
